keras_nlp/layers/modeling/lora_dense.py

```diff
@@ -0,0 +1,230 @@
+    parameters, and builds a low-rank decomposed update to sum with the original
+    `inner_dense` output. These update parameters can be merged back into the
+    `inner_dense` kernel by calling `merge_weights()`.
+
+    Args:
+        inner_dense: A `keras.layers.Dense` or `keras.layers.EinsumDense`.
+            The inner dense layer to freeze and wrap with the `LoraDense`
+            layer. Note that for `EinsumDense` layers, the einsum equation must
+            represent a dense transformation on the last axis of the input,
+            though adding new axes to the output (e.g. a multi-head axis) is
+            allowed.
+        rank: int The inner rank of the decomposed dense transformation. The
+            lower this number, the less trainable parameters the layer will
+            have.
+        alpha: float. A constant value used for scaling the lora update. The
+            lora update to the original dense transformation will be scaled by
+            `alpha / rank`.
+        lora_a_initializer: The initializer to use for the inner projection
+            from layer inputs to the inner `rank` intermediate outputs.
+        freeze_kernel: If true, the kernel of the inner dense layer will have
+            `trainable` set to False.
+        freeze_bais: If true, the kernel of the inner dense layer will have
+            `trainable` set to False.
+        **kwargs: other keyword arguments.
+
+    Examples:
+
+    Wrap a `Dense` layer.
+    ```python
+    batch_size, feature_size = 4, 16
+    rank = 4
+    inputs = np.random.uniform(size=(batch_size, feature_size))
+    inner_dense = keras.layers.Dense(feature_size)
+    lora_dense = keras_nlp.layers.LoraDense(inner_dense, rank=4)
+    # Output with inner dense begins equal.
+    assert np.allclose(inner_dense(inputs), lora_dense(inputs))
+
+    # Add some random updates to the lora parameters.
+    lora_dense.lora_a.assign(np.random.uniform(size=(feature_size, rank)))
+    lora_dense.lora_b.assign(np.random.uniform(size=(rank, feature_size)))
+    assert not np.allclose(inner_dense(inputs), lora_dense(inputs))
+
+    # Merge the lora dense and output
+    lora_dense.merge_weights()
+    assert np.allclose(inner_dense(inputs), lora_dense(inputs))
+    ```
+
+    Wrap an `EinsumDense` layer with a multi-head projection.
+    ```python
+    batch_size, sequence_length, feature_size = 4, 10, 16
+    num_heads = 2
+    rank = 4
+    inputs = np.random.uniform(size=(batch_size, sequence_length, feature_size))
+    inner_dense = keras.layers.EinsumDense(
+        "abc,cde->abde",
+        output_shape=(sequence_length, num_heads, feature_size // num_heads),
+    )
+    lora_dense = keras_nlp.layers.LoraDense(inner_dense, rank=4)
+    # Output shape (4, 10, 2, 8)
+    lora_dense(inputs)
+    ```
+    """
+
+    def __init__(
+        self,
+        inner_dense,
+        rank=8,
+        alpha=32.0,
+        lora_a_initializer="variance_scaling",
```

Comment:
Does it make sense to include "lora" as a prefix for this variable name, given that it's already a LoraDense layer/

Reply:
I think it's okay, because the two dense layers in LoRA are called "lora_A" and "lora_B" (in the [official code](https://github.com/microsoft/LoRA/blob/main/loralib/layers.py#L48)). Calling this `a_initializer` would look weird :P.

Reply:
Yeah, I wanted to avoid `layer.a` and `layer.b`. We could come up with our own names, `layer.inner_kernel_update` and `layer.outer_kernel_update`, but I suspect `lora_a` and `lora_b` will be more recognizable to people.

Reply:
Sounds fine to me.